atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/cleaner/KeyValueServicePuncherStore.java

```diff
@@ -105,59 +134,83 @@ public boolean isInitialized() {
-    public long getMillisForTimestamp(long timestamp) {
-        return getMillisForTimestamp(keyValueService, timestamp);
+    private Optional<Long> getHistoric(long timeMillis, int numRowsToCheck) {
+        List<byte[]> rows = LongStream.rangeClosed(1, numRowsToCheck)
+                .map(i -> timeMillis - i * MILLIS_IN_TWO_WEEKS)
+                .mapToObj(KeyValueServicePuncherStore::row)
+                .collect(Collectors.toList());
+        Map<byte[], RowColumnRangeIterator> itMap = keyValueService.getRowsColumnRange(
+                AtlasDbConstants.PUNCH_TABLE,
+                rows,
+                BatchColumnRangeSelection.create(null, null, 1),
+                Long.MAX_VALUE);
+
+        return rows.stream()
+                .map(itMap::get)
+                .filter(Iterator::hasNext)
+                .map(Iterator::next)
+                .map(Map.Entry::getValue)
+                .map(KeyValueServicePuncherStore::getAtlasTimestamp)
+                .findFirst();
+    }
+
+    private static long getAtlasTimestamp(Value value) {
+        return EncodingUtils.decodeUnsignedVarLong(value.getContents());
     }
 
+    // Relies on a full table scan, but should only be used in a cli.
     public static long getMillisForTimestamp(KeyValueService kvs, long timestamp) {
         long timestampExclusive = timestamp + 1;
-        // punch table is keyed by the real value we're trying to find so we have to do a whole table
-        // scan, which is fine because this table should be really small
-        byte[] startRow = EncodingUtils.encodeUnsignedVarLong(Long.MAX_VALUE);
-        EncodingUtils.flipAllBitsInPlace(startRow);
-        RangeRequest rangeRequest =
-                RangeRequest.builder().startRowInclusive(startRow).batchHint(1).build();
-        ClosableIterator<RowResult<Value>> result =
-                kvs.getRange(AtlasDbConstants.PUNCH_TABLE, rangeRequest, timestampExclusive);
-
-        try {
-            if (result.hasNext()) {
-                byte[] encodedMillis = result.next().getRowName();
-                EncodingUtils.flipAllBitsInPlace(encodedMillis);
-                return EncodingUtils.decodeUnsignedVarLong(encodedMillis);
-            } else {
-                return 0L;
-            }
-        } finally {
-            result.close();
+        RangeRequest rangeRequest = RangeRequest.builder().batchHint(1).build();
+        try (ClosableIterator<RowResult<Value>> result =
+                kvs.getRange(AtlasDbConstants.PUNCH_TABLE, rangeRequest, timestampExclusive)) {
+            // With Guava 21.0, can do Streams.stream(result), Streams.stream(row.getCells())
+            return StreamSupport.stream(Spliterators.spliteratorUnknownSize(result, 0), false)
+                    .flatMap(row -> StreamSupport.stream(row.getCells().spliterator(), false))
+                    .map(Map.Entry::getKey)
+                    .mapToLong(KeyValueServicePuncherStore::timestamp)
+                    .max()
```

Comment:
Can we do `max` and then `mapToLong(KeyValueServicePuncherStore::timestamp)`? This seems like is going to do 1 KVS call per atlas timestamp, and then take the max.

Not sure if it's needed, since this is also an approximation of an approximation...